How to use a functor to replace the lambda one?

Hello,
Thank you for this great sqlite wrapper! It's very convenient.
While using the wrapper is delightful, a problem about eliminating similar code occurs.

The functor:
```c++
struct ItemResultHandler {
    explicit ItemResultHandler(QVariantMap& map_) : map(map_) {}
    void operator() ( int id, std::string name, double price, int qty ) {
         map.insert("id", id);
         map.insert("name", QString::fromStdString(name));   
         map.insert("price", price);
         map.insert("qty", qty);
    }
    QVariantMap& map;
};
```
The code:
```c++
QVariantMap getItemById(int id) {
  QVariantMap map;
  database db(dbfile);
  db << "select id,name,price,qty from items where id=?"
     << id 
     >> ItemResultHandler(map);
  return map;
}
```
It has compiling errors:
```
../../myapp/src/utility/function_traits.h:12: error: type 'void (ItemResultHandler::*)(std::__1::basic_string<char>, std::__1::basic_string<char>, int, int)' cannot be used prior to '::' because it has no members
        decltype(&Function::operator())
                  ^
```
```
 ../../myapp/src/sqlite_modern_cpp.h:218: error: incomplete definition of type 'sqlite::utility::function_traits<ItemResultHandler>'
                        binder<traits::arity>::run(*this, func);
                               ~~~~~~^~
```
I have posted a question in SO. Could you please have a look at it? Thank you in advance.
http://stackoverflow.com/questions/30387896/how-to-eliminate-multiple-similar-sql-queries-with-one-resultset-hander

--- a/hdr/sqlite_modern_cpp/utility/function_traits.h
+++ b/hdr/sqlite_modern_cpp/utility/function_traits.h
@@ -31,5 +31,26 @@ namespace sqlite {
 			static const std::size_t arity = sizeof...(Arguments);
 		};
 
+    /* support the non-const operator ()
+     * this will work with user defined functors */
+		template <
+			typename    ClassType,
+			typename    ReturnType,
+			typename... Arguments
+		>
+		struct function_traits<
+			ReturnType(ClassType::*)(Arguments...)
+		> {
+			typedef ReturnType result_type;
+
+			template <std::size_t Index>
+			using argument = typename std::tuple_element<
+				Index,
+				std::tuple<Arguments...>
+			>::type;
+
+			static const std::size_t arity = sizeof...(Arguments);
+		};
+
 	}
 }

--- a/tests/functors.cc
+++ b/tests/functors.cc
@@ -19,17 +19,17 @@ struct tbl_functor {
 int main() {
 
   try {
-		database db(":memory:");
-		db << "CREATE TABLE tbl (id integer, name string);";
-		db << "INSERT INTO tbl VALUES (?, ?);" << 1 << "hello";
-		db << "INSERT INTO tbl VALUES (?, ?);" << 2 << "world";
+    database db(":memory:");
+    db << "CREATE TABLE tbl (id integer, name string);";
+    db << "INSERT INTO tbl VALUES (?, ?);" << 1 << "hello";
+    db << "INSERT INTO tbl VALUES (?, ?);" << 2 << "world";
 
     vector<pair<int,string> > vec;
     db << "select id,name from tbl;" >> tbl_functor(vec);
 
     if(vec.size() != 2) {
-			cout << "Bad result on line " << __LINE__ << endl;
-			exit(EXIT_FAILURE);
+      cout << "Bad result on line " << __LINE__ << endl;
+      exit(EXIT_FAILURE);
     }
 
     vec.clear();
@@ -38,20 +38,20 @@ int main() {
     db << "select id,name from tbl;" >> functor;
 
     if(vec.size() != 2 || vec[0].first != 1 || vec[0].second != "hello") {
-			cout << "Bad result on line " << __LINE__ << endl;
-			exit(EXIT_FAILURE);
+      cout << "Bad result on line " << __LINE__ << endl;
+      exit(EXIT_FAILURE);
     }
 
   }
-	catch(sqlite_exception e) {
-		cout << "Unexpected error " << e.what() << endl;
-		exit(EXIT_FAILURE);
-	}
-	catch(...) {
-		cout << "Unknown error\n";
-		exit(EXIT_FAILURE);
-	}
+  catch(sqlite_exception e) {
+    cout << "Unexpected error " << e.what() << endl;
+    exit(EXIT_FAILURE);
+  }
+  catch(...) {
+    cout << "Unknown error\n";
+    exit(EXIT_FAILURE);
+  }
 
-	cout << "OK\n";
-	exit(EXIT_SUCCESS);
+  cout << "OK\n";
+  exit(EXIT_SUCCESS);
 }

--- a/tests/functors.cc
+++ b/tests/functors.cc
@@ -1,0 +1,57 @@
+#include <iostream>
+#include <vector>
+#include <algorithm>
+#include <string>
+#include <sqlite_modern_cpp.h>
+
+using namespace sqlite;
+using namespace std;
+
+struct tbl_functor {
+  explicit tbl_functor(vector<pair<int, string> > &vec_) : vec(vec_) { }
+
+  void operator() ( int id, string name) {
+    vec.push_back(make_pair(id, move(name)));
+  }
+  vector<pair<int,string> > &vec;
+};
+
+int main() {
+
+  try {
+		database db(":memory:");
+		db << "CREATE TABLE tbl (id integer, name string);";
+		db << "INSERT INTO tbl VALUES (?, ?);" << 1 << "hello";
+		db << "INSERT INTO tbl VALUES (?, ?);" << 2 << "world";
+
+    vector<pair<int,string> > vec;
+    db << "select id,name from tbl;" >> tbl_functor(vec);
+
+    if(vec.size() != 2) {
+			cout << "Bad result on line " << __LINE__ << endl;
+			exit(EXIT_FAILURE);
+    }
+
+    vec.clear();
+
+    tbl_functor functor(vec);
+    db << "select id,name from tbl;" >> functor;
+
+    if(vec.size() != 2 || vec[0].first != 1 || vec[0].second != "hello") {
+			cout << "Bad result on line " << __LINE__ << endl;
+			exit(EXIT_FAILURE);
+    }
+
+  }
+	catch(sqlite_exception e) {
+		cout << "Unexpected error " << e.what() << endl;
+		exit(EXIT_FAILURE);
+	}
+	catch(...) {
+		cout << "Unknown error\n";
+		exit(EXIT_FAILURE);
+	}
+
+	cout << "OK\n";
+	exit(EXIT_SUCCESS);
+}

